### PR TITLE
[Link Event Damping] Add Abstract SelectEventHandler class to handle link event damping timer events.

### PR DESCRIPTION
**Pull request by Ashish1805**
- Handler in the class will be used to handle the timer event of link event dampers.

HLD: sonic-net/SONiC#1071

### CODE REVIEW

File: syncd/SelectableEventHandler.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include "swss/logger.h"
+#include "swss/sal.h"
+
+namespace syncd
+{
+
```
Comment (kcudnik) [inline syncd/SelectableEventHandler.h:8]: empty line not needed

File: syncd/SelectableEventHandler.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    // This class implements handler for Selectable events.
+    class SelectableEventHandler {
```
Comment (kcudnik) [inline syncd/SelectableEventHandler.h:10]: add { at the next line

File: syncd/SelectableEventHandler.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        public:
+
+            virtual ~SelectableEventHandler()
+            {
+              SWSS_LOG_ENTER();
+            }
```
Comment (kcudnik) [inline syncd/SelectableEventHandler.h:16]: if its not a template, move code to cpp file

Comment (Ashish1805) [reply syncd/SelectableEventHandler.h:16]: Do we see any issue with keeping the destructor here in .h file?

**Comment by Ashish1805** [comment]
Adding @Junchao-Mellanox for review.

File: syncd/SelectableEventHandler.h (hunk 3 above)
Comment (kcudnik) [reply syncd/SelectableEventHandler.h:16]: Usually when code is not template we prefer to move cpp code to cpp file, and also add empty intentionally or = default

File: syncd/SelectableEventHandler.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+            virtual void handleSelectableEvent() = 0;
+
+        protected:
+
+            SelectableEventHandler()
```
Comment (Junchao-Mellanox) [inline syncd/SelectableEventHandler.h:22]: do we really need this?

File: syncd/SelectableEventHandler.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include "swss/logger.h"
+#include "swss/sal.h"
```
Comment (Junchao-Mellanox) [inline syncd/SelectableEventHandler.h:4]: unused include?

File: syncd/SelectableEventHandler.h (hunk 4 above)
Comment (kcudnik) [reply syncd/SelectableEventHandler.h:22]: Why constructor is protected ?

**Review by kcudnik — CHANGES_REQUESTED** [review]

**Review by Junchao-Mellanox — APPROVED** [review]

**Comment by Ashish1805** [comment]
@kcudnik Please let me know if you have any other comments/concerns. If everything looks good, can you please merge the PR? Thank you.

**Comment by kcudnik** [comment]
You did not answered my questions about protected constructors

**Review by kcudnik — APPROVED** [review]

**Comment by Ashish1805** [comment]
> You did not answered my questions about protected constructors

In another comment reply, I had added the explanation:

Strictly speaking constructor is not needed, but constructor was added as protected just to make it clear to clients that they cannot directly instantiate an object of the class.

Agree, this class still be uninstantiable because of the pure virtual method, but to put the emphasize on not trying to directly instantiate this class, protected constructor was kept.

I am inclined to keep constructor as:
protected:
SelectableEventHandler() = default;

Thank you.

**Pull request merged**